products/llbuild-analyze/Sources/CriticalPathTool.swift

```diff
@@ -272,3 +291,119 @@ struct EncodableRuleResult: Encodable {
         try container.encode(deps, forKey: .dependencies)
     }
 }
+
+
+protocol GraphVizNode {
+    var graphVizName: String { get }
+}
+
+extension BuildKey.Command: GraphVizNode {
+    var graphVizName: String {
+        "Command:\(self.name)"
+    }
+}
+
+extension BuildKey.CustomTask: GraphVizNode {
+    var graphVizName: String {
+        "CustomTask:\(self.name)"
+    }
+}
+
+extension BuildKey.DirectoryContents: GraphVizNode {
+    var graphVizName: String {
+        "DirectoryContents:\(self.path)"
+    }
+}
+
+extension BuildKey.FilteredDirectoryContents: GraphVizNode {
+    var graphVizName: String {
+        "FilteredDirectoryContents:\(self.path)"
+    }
+}
+
+extension BuildKey.DirectoryTreeSignature: GraphVizNode {
+    var graphVizName: String {
+        "DirectoryTreeSignature:\(self.path)"
+    }
+}
+
+extension BuildKey.DirectoryTreeStructureSignature: GraphVizNode {
+    var graphVizName: String {
+        "DirectoryTreeStructureSignature:\(self.path)"
+    }
+}
+
+extension BuildKey.Node: GraphVizNode {
+    var graphVizName: String {
+        "Node:\(self.path)"
+    }
+}
+
+extension BuildKey.Stat: GraphVizNode {
+    var graphVizName: String {
+        "Stat:\(self.path)"
+    }
+}
+
+extension BuildKey.Target: GraphVizNode {
+    var graphVizName: String {
+        "Target:\(self.name)"
+    }
+}
+
+/// Struct to represent a directed edge in GraphViz from a -> b.
+/// `hash()` and `==` only take the both edges into account, not
+/// `isCritical`, so the graph can be represented as a `Set<DirectedEdge>`
+/// and gurantee that there is only one edge between two verticies.
+struct DirectedEdge: Hashable, Equatable {
+    /// Source `BuildKey`
+    let a: BuildKey
+
+    /// Destination `BuildKey`
+    let b: BuildKey
+
+    /// Flag if the edge is on critical build path.
+    var isCritical: Bool = false
+
+    static func == (lhs: Self, rhs: Self) -> Bool {
+        lhs.a == rhs.a && lhs.b == rhs.b
+    }
+
+    func hash(into hasher: inout Hasher) {
+        a.hash(into: &hasher)
+        b.hash(into: &hasher)
+    }
+
+    /// Style attributes for the edge.
+    private var style: String {
+        if isCritical {
+            return "[style=bold]"
+        }
+        return ""
+    }
+
+    /// GraphViz representation of the Edge.
+    var graphVizString: String {
+        guard let a = a as? GraphVizNode, let b = b as? GraphVizNode else {
+            fatalError("Both edges need to conform to GraphvizNode")
+        }
+        return "\t\"\(a.graphVizName)\" -> \"\(b.graphVizName)\"\(style)\n"
+    }
+
+}
+
+private protocol CommandLineArgumentChoices: CaseIterable, RawRepresentable where RawValue == String {
```

Comment:
Would it make sense to define the protocol without type constraints and have the extension use them so that other types could implement this as well? Would probably need a deeper implementation and a more sophisticated help description system to make sense.

Reply:
@BenchR267 I think that makes sense,  however, I am not sure if that should be in scope for SR-11730. Should I open a new ticket at bugs.swift.org?

Reply:
Yeah, that makes sense. Yes, please. Thanks!